asmcomp/loongarch64/proc.ml

```diff
@@ -0,0 +1,305 @@
+# 2 "asmcomp/loongarch64/proc.ml"
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*                yala <zhaojunchao@loongson.cn>                          *)
+(*                                                                        *)
+(*               Copyright © 2008-2023 LOONGSON                           *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(* Description of the LoongArch *)
+
+open Misc
+open Cmm
+open Reg
+open Arch
+open Mach
+
+(* Registers available for register allocation *)
+
+(* Integer register map
+   --------------------
+
+    zero                   always zero
+    ra                     return address
+    sp, gp, tp             stack pointer, global pointer, thread pointer
+    a0-a7        0-7       arguments/results
+    s2-s6        8-12      arguments/results (preserved by C)
+    t2-t6        13-17     temporary
+    s0           18        general purpose (preserved by C)
+    t0, t1       19-20     temporaries (used by call veneers)
+    s1           21        trap pointer (preserved by C)
+    s7           22        allocation pointer (preserved by C)
+    s8           23        domain pointer (preserved by C)
+
+  Floating-point register map
+  ---------------------------
+
+    ft0-ft7    100-107     temporary
+    fs0-fs1    108-109     general purpose (preserved by C)
+    fa0-fa7    110-117     arguments/results
+    fs2-fs7    118-123     general purpose (preserved by C)
+    ft8-f15    124-131     temporary
+
+  Additional notes
+  ----------------
+
+    - t1 is used by the code generator, so not available for register
+      allocation.
+
+    - t0-t6 may be used by PLT stubs, so should not be used to pass
+      arguments and may be clobbered by [Ialloc] in the presence of dynamic
+      linking.
+*)
+
+let int_reg_name =
+    [|"$a0"; "$a1"; "$a2"; "$a3"; "$a4"; "$a5"; "$a6"; "$a7";  (* 0- 7 *)
+      "$s2"; "$s3"; "$s4"; "$s5"; "$s6";                       (* 8-12 *)
+      "$t2"; "$t3"; "$t4"; "$t5"; "$t6"; "$t7"; "$t8";         (* 13-19 *)
+      "$s0";                                                   (* 20 *)
+      "$t0"; "$t1";                                            (* 21-22 *)
+      "$s1"; "$s7"; "$s8";                                     (* 23-25 *)
```

Comment:
This is different than [the official register convention](https://github.com/loongson/la-abi-specs/blob/v2.30/lapcs.adoc#general-purpose-registers), any particular reason why the deviation is necessary?

Reply:
这个ABI 在编译ocaml代码时生效，在ocaml内部遵守。
This ABI takes effect when the ocaml code is compiled and is observed internally by ocaml.

Reply:
> 这个ABI 在编译ocaml代码时生效，在ocaml内部遵守。 This ABI takes effect when the ocaml code is compiled and is observed internally by ocaml.
> 
> 当ocmal要调用c代码时 runtime/loongarch64.S caml_c_call函数会对调用约定进行转换

It could lighten maintenance burden (think of extra things to care for) and reduce context-switching costs if the OCaml ABI could adhere to the platform ABI as closely as possible. What do you think?

Reply:
Also it's kind of important that the reserved `$r21` register should not get used.

Reply:
在ocaml上下文中，需要占用几个固定的寄存器来保存一些数据指针，这是由ocaml的整体设计决定。ocaml上下文中遵循他自己的这套寄存器使用约定并没有什么问题，其他架构也是这样做的。
In the ocaml context, the need to occupy a few fixed registers to hold some data pointers is dictated by the overall design of ocaml. There is nothing wrong with following his own set of register usage conventions in the ocaml context, as other architectures do as well.
```
  "$t0"; "$t1";                                            (* 21-22 *)
```
这里的数字21是指t0寄存器在这个数组中的编号，并不是指$21 。
The number 21 here refers to the number of the t0 register in this array, not to $21 .